### PR TITLE
Fused unrolled backwards pass specialized on downstream counter

### PR DESCRIPTION
**Pull request by maleadt**
Supersede #20 #21 #22

### CODE REVIEW

**Comment by maleadt** [comment]
```
| environment | precomputed layers? | SoA enabled? | size | forward time | backward time |
| -----------:|:-------------------:|:------------:|:----:|:------------:|:-------------:|
|        :gpu |        false        |     true     | 512  |  188.05 us   |   265.13 us   |
|        :gpu |        false        |     true     | 1024 |  288.43 us   |   342.17 us   |
|        :gpu |        false        |     true     | 2048 |  690.16 us   |   661.83 us   |
|        :gpu |        true         |     true     | 512  |  128.73 us   |   222.40 us   |
|        :gpu |        true         |     true     | 1024 |  166.90 us   |   254.37 us   |
|        :gpu |        true         |     true     | 2048 |  377.22 us   |   391.59 us   |
```

File: src/ad/primitives.jl [hunk 1]
```diff
@@ -165,35 +165,108 @@ end
...
+    body = Expr(:block)
+    for i in 1:N
+        if inputs.parameters[i] <: Variable
+            input = :(inputs[$i])
+            push!(body.args, quote
+                if $input.downstreams > 1
+                    broadcast!(multivariable_backprop_partial, deriv($input), deriv($input),
+                               output_duals, $i, output_derivs)
+                else
+                    broadcast!(backprop_partial, deriv($input), output_duals, $i,
```
Comment (vchuravy) [inline src/ad/primitives.jl:213]: Ok so instead of specialising on `Val` we are now relying on inlining and const prop?

Comment (maleadt) [reply src/ad/primitives.jl:213]: yeah.

**Comment by maleadt** [comment]
Going to merge this in order to cut down on open PRs.

**Pull request merged**